src/posix-socket.c

```diff
@@ -0,0 +1,612 @@
+    if(s->closed){
+        return JS_ThrowInternalError(ctx, "Socket closed");
+    }
+    int flags;
+    size_t sz;
+    uint8_t* buf = JS_GetUint8Array(ctx, &sz, argv[0]);
+    CHECK_ARG_RET(ctx, buf, 0, "Uint8Array");
+    CHECK_ARG_RET(ctx, !JS_ToUint32(ctx, (unsigned*)&flags, argv[1]), 1, "positive integer");
+    size_t addrsz;
+    struct sockaddr* addr = JS_GetUint8Array(ctx, &addrsz, argv[2]);
+    CHECK_ARG_RET(ctx, addr != NULL, 2, "Uint8Array");
+    int ret = sendto(s->sock, buf, sz, flags, addr, addrsz);
+    RET_THROW_ERRNO(ctx, ret >= 0);
+    return JS_NewUint32(ctx, ret);
+}
+*/
+
+static JSValue tjs_sock_sendmsg(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
+    // this: PosixSocket
+    // args: Uint8Array|undefined addr, Uint8Array|undefined control, int flags, Uint8Array ...data
+    tjs_sock_t* s = JS_GetOpaque(this_val, tjs_sock_classid);
+    CHECK_ARG_RET(ctx, s, -1, tjs_sock_class.class_name);
+    if(s->closed){
+        return JS_ThrowInternalError(ctx, "Socket closed");
+    }
+    struct msghdr msg;
+    
+    if(argc < 4){
+        return JS_ThrowInternalError(ctx, "expected at least 4 arguments");
+    }
+
+    if(!JS_IsUndefined(argv[0])){
+        size_t addrsz;
+        struct sockaddr* addr = (struct sockaddr*)JS_GetUint8Array(ctx, &addrsz, argv[0]);
+        CHECK_ARG_RET(ctx, addr != NULL, 0, "Uint8Array");
+        msg.msg_name = addr;
+        msg.msg_namelen = addrsz;
+    }else{
+        msg.msg_name = NULL;
+        msg.msg_namelen = 0;
+    }
+    if(!JS_IsUndefined(argv[1])){
+        size_t ctrlsz;
+        uint8_t* ctrl = JS_GetUint8Array(ctx, &ctrlsz, argv[1]);
+        CHECK_ARG_RET(ctx, ctrl != NULL, 1, "Uint8Array");
+        msg.msg_control = ctrl;
+        msg.msg_controllen = ctrlsz;
+    }else{
+        msg.msg_control = NULL;
+        msg.msg_controllen = 0;
+    }
+    int flags;
+    CHECK_ARG_RET(ctx, !JS_ToUint32(ctx, (unsigned*)&flags, argv[1]), 1, "positive integer");
+    msg.msg_flags = flags;
+
+    msg.msg_iovlen = (argc - 3);
+    msg.msg_iov = js_malloc(ctx, sizeof(struct iovec) * msg.msg_iovlen);
+
+    for(int i = 0; i < msg.msg_iovlen; i++){
+        size_t sz;
+        uint8_t* buf = JS_GetUint8Array(ctx, &sz, argv[i+3]);
+        if(buf == NULL){
+            js_free(ctx, msg.msg_iov);
+            return THROW_ARG_ERR(ctx, i+3, "Uint8Array");
+        }
+        msg.msg_iov[i].iov_base = buf;
+        msg.msg_iov[i].iov_len = sz;
+    }
+    int ret = sendmsg(s->sock, &msg, flags);
+    js_free(ctx, msg.msg_iov);
+    RET_THROW_ERRNO(ctx, ret >= 0);
+    return JS_NewUint32(ctx, ret);
+}
+
+
+
+static void tjs_sock_uv_poll_cb(uv_poll_t* handle, int status, int events){
+	tjs_sock_t* s = uv_handle_get_data((uv_handle_t*)handle);
+	JSValueConst args[] = {
+		JS_NewInt32(s->jsctx, status),
+		JS_NewInt32(s->jsctx, events)
+	};
+	s->in_cb = true;
+	JSValue ret = JS_Call(s->jsctx, s->callback, s->this, countof(args), args);
```

Comment:
I think you need to dup the callback here to avoid problems if you call close in the callback.

Reply:
close in the callback shouldn't happen (libuv doesn't like that). 
So I added a check (like in poll_stop).